### ISSUE TITLE
🔧 chore: llama.cpp pin 上げの検討（b8694 → 現行）— QAT shared-KV 対応の可否を spike で判断

### ISSUE THREAD

**Issue by tyabu12** [tyabu12/pastura#1415]
## Summary

Investigate bumping the `mattt/llama.swift` pin from the current `exact: "2.8694.0"`
(llama.cpp **b8694**) to a current release, and decide go/no-go. Scoped as a
**discardable spike**, not a production bump — the motivating upside is currently
unmeasured (see § Motivation).

Latest available at filing: **2.10327.0** (llama.cpp **b10327**) — roughly **1,633
builds** of drift.

## Motivation

Gemma 4 E2B **QAT** GGUFs (both Google's official
`google/gemma-4-E2B-it-qat-q4_0-gguf` and unsloth's
`unsloth/gemma-4-E2B-it-qat-GGUF`) ship a **shared-KV tail-layer** layout: 541
tensors, omitting `attn_k` / `attn_v` / `attn_k_norm` for layers 15–34 (60 tensors)
that the shipped non-QAT `unsloth/gemma-4-E2B-it-GGUF` Q4_K_M materializes (601
tensors). Pinned b8694 has no shared-KV tail-layer loader and aborts:

```
llama_model_load: error loading model: missing tensor 'blk.15.attn_k.weight'
```

This blocked a `/model-eval` Gate 1 run at model load, 6/6 cells, zero inferences.

**Verified first-hand** via a standalone throwaway SwiftPM package (the repo pin was
not touched), rather than taken from an upstream thread:

| llama.cpp | Google QAT Q4_0 (541 tensors) | incumbent Q4_K_M (601 tensors) |
|---|---|---|
| b8694 (current pin) | `missing tensor 'blk.15.attn_k.weight'` | LOADED OK |
| **b10327** (latest) | **LOADED OK** | **LOADED OK** |

`llama_vocab_is_eog(106)` stays `true` on both builds for both files, so `<turn|>`
still terminates generation.

So a pin bump *would* unlock the QAT candidate. **But there is no evidence yet that
QAT improves Pastura's output** — the blocked eval produced zero tokens. That makes
"bump the pin to get QAT" a spend against a speculative gain, which is why the
sequencing below measures the delta before committing to the bump.

## Why this is not a routine dependency bump

Compile-time breakage (e.g. historical `llama_token_*` → `llama_vocab_*` renames) is
the **loud, tractable** class. The quieter risks:

1. **Grammar-sampler coupling.** Pastura's deepest llama.cpp coupling sits in the
   highest-churn subsystem: `LlamaCppService+GrammarSample.swift` (two-pass
   grammar-constrained sampling, #751), `SafeSampler` (Obj-C++ exception bridge,
   #253 / #371 / #463), `LlamaCppService+DrySampler.swift` (custom DRY, ring
   buffer), plus the EOG-skip invariant on grammar accept (#253). These encode
   *observed b8694 behavior*, not documented API contracts. A behavioral change
   compiles clean and surfaces at runtime as empty output or `GGML_ASSERT` →
   SIGABRT — the failure family ADR-011 spent three diagnostic rounds on.
2. **Eval-baseline re-basing.** `docs/models/eval-log.md` verdicts, the gitignored
   `data/models/eval-digest.md` scores, and the `samplerGrammarResample` position-0
   fragility metric were all measured on b8694. Post-bump, cross-model comparisons
   (including the Sarashina NO-GO) stop being like-for-like.
3. **Device-only risk.** Metal behavior, the ~8 GB RAM headroom estimate
   (ADR-002 §4), and app binary size are not observable from the Mac harness — the
   reason ADR-011 Gate 2 exists at all.
4. **Is `SafeSampler` still load-bearing?** Upstream `ggml-org/llama.cpp#21017` was
   closed as not planned, so probably yes — but re-verify rather than assume, and
   check the workaround is still *sufficient* as well as still *necessary*.

## Proposed approach — discardable spike first

Do **not** open a pin-bump PR as step 1.

1. In a throwaway worktree, bump the pin locally and enumerate the compile-time
   breakage surface (build the app target + `swift build` for the harness).
2. Re-verify the sampler invariants explicitly, each against a negative control:
   grammar two-pass path, `SafeSampler` catch behavior, EOG-skip on grammar accept,
   DRY sampler.
3. Run the `/model-eval` 6-cell battery for **both incumbents** (Gemma 4 E2B
   Q4_K_M, Qwen 3 4B) on the bumped pin and diff against the b8694 baseline —
   this is the regression signal *and* re-bases the metric.
4. Run the battery for the Google QAT candidate to finally measure the QAT delta.
5. **Decide**: if the QAT delta is not materially better than the incumbent, close
   this as no-go and record it — the drift alone is not a reason to bump.
6. If go: land the bump via `/orchestrate` with an ADR-002 amendment (§ Pin Strategy
   documents the verification bar: loads without crash, JSON parses, no hangs over
   10+ inferences) + ADR-011 P4 catalog regression + real-device QA.

## Acceptance criteria

- [ ] Compile-time breakage surface enumerated for app + harness targets
- [ ] Sampler invariants (#253 / #371 / #463 / #751) re-verified on the new build,
      each with a negative control that demonstrably fires
- [ ] Battery re-run for both incumbents on the new pin, diffed vs the b8694 baseline
- [ ] QAT delta measured (or explicitly recorded as still-unmeasurable, with cause)
- [ ] Written go/no-go recommendation; if go, the ADR-002 amendment plan
- [ ] Spike worktree discarded — no pin change on `main` from this issue

## Not `agent-ready`

Harness-gated stochastic judgment loop (same rationale as the `/model-eval`
positioning), **plus** it ends in a dependency decision requiring maintainer
approval per CLAUDE.md § Confirmation Policy. Run interactively.

## References

- Blocked Gate 1 run: 2026-08-08, `data/models/eval-digest.md`
  § `2026-08-08 — gemma-4-e2b-q4-k-m (gemma-4-E2B_q4_0-it)` (gitignored, local)
- ADR-002 § Pin Strategy, § 3 Model Specification
- ADR-011 § Decision item 2 (five hard prerequisites)
- Intake queue: #979
- Upstream context: [ikawrakow/ik_llama.cpp#1928](https://github.com/ikawrakow/ik_llama.cpp/issues/1928)

**Comment by tyabu12** [comment]
## Implementation Plan

Residual documentation drift left over from this issue's pin bump (#1489). Three
prose corrections; no code, no behaviour change.

- [x] 1. 🎭 Split `docs/models/eval-log.md` §2026-08-08's two P3–P5 statements by provenance (`docs/models/eval-log.md`) — `:490-491` was authored by `8bf34195` (2026-08-16), *after* the 2026-08-15 device PASS it contradicts, so it is born-wrong and gets rewritten to route to §2026-08-13 rather than restate a status; `:506` was authored by `1adb2bd1` (2026-08-13) and was true when written, so it gets the file's own annotate-and-scope idiom (cf. `:316`) naming this entry's own Google QAT Q4_0 candidate.
- [x] 2. 🎭 Correct the `common_sampler_sample` characterization to the measured cross-tag diff, and sweep its mirror (`docs/decisions/ADR-002.md`, `.claude/rules/engine.md`) — the changed region is the backend-sampler early return's internals **plus** the `set_logits` hoist above the block; `rbudget` in both passes is **not** new at b10327 (byte-identical at b8694), so the "one new divergence" reading is retracted and Pastura's missing counterpart is recorded as pre-existing. Citations become filename-qualified and one-line per ADR-002 verification-bar item 6 (the two sibling bullets in the same paragraph already wrap, and are swept with it). Carries a dated correction clause so a reader arriving from #1489's body sees the reconciliation. `.claude/rules/engine.md:423-426` mirrors the old "changed **only** inside a backend-sampler early return" wording and is fixed in the same commit.
- [ ] 3. 🎭 Add one clause to ADR-002 verification-bar item 2 (`docs/decisions/ADR-002.md`) — "re-read the upstream body" is what let a stage present at b8694 be recorded as new at b10327; the bar now says to diff the body **across both tags**. One clause, not a new numbered item.

## Verification notes

- The load-bearing claim (the b8694↔b10327 body diff) exists in **no in-repo artifact** — the
  resolved xcframework ships headers only. Measured by fetching
  `raw.githubusercontent.com/ggml-org/llama.cpp/<tag>/common/sampling.cpp` for both tags and
  running `diff -u` on the two function bodies. Extents: `common/sampling.cpp:574-656` (b10327),
  `common/sampling.cpp:526-606` (b8694).
- Population for item 1 re-derived from the authoring commit, not only from a `P3–P5` token grep.

## Metadata
- **Type**: docs
- **Branch**: `docs/1415-residual-doc-drift`
- **Reviewer**: Opus (reason: touches docs/decisions/ADR-*.md and .claude/rules/** — Opus-required twice over)
- **Session**: Opus (reason: 1–2 sentence prose edits whose measured evidence is already held in the orchestrator's context; delegation overhead exceeds the implementation)

**Comment by tyabu12** [comment]
## Closed — the spike's NO-GO was reversed, the bump landed, and the residual drift it left is swept

Auto-closed by #1491. Recording the arc, because the verdict changed twice and the comment thread
reads misleadingly if only its middle is skimmed.

### What this issue actually decided

| Question it was chartered to answer | Answer |
|---|---|
| Does a wrapper release exist at a shared-KV-capable build? | Yes — `2.10327.0` / llama.cpp b10327 |
| How expensive is the bump? | Two call sites, no measured regression on either incumbent |
| Does QAT justify it? | **Not for the file first measured** — `google/…-qat-q4_0-gguf` scored 103 vs 113 and is +7.8 % larger. NO-GO as written, and it stands for that file |
| …for the QAT *family*? | **Yes, on a different member.** `unsloth/gemma-4-E2B-it-qat-GGUF` → `UD-Q4_K_XL`: −15.7 % size, no measured quality regression against a same-session control |
| QAT-Mobile `UD-Q2_K_XL`, the −30 % candidate? | Unavailable at any published build — llama.cpp ships no `TQ`-family Metal kernels. Tracked separately in #1416 |

The scope correction is the part worth carrying forward: the first NO-GO was a fact about **one
export**, not about QAT. Two of the three legs it stood on (quality delta, size direction) reversed
when the sibling file was measured — the pin-cost finding is the only leg that generalised.

### Shipped

- **#1484** — DRY sampler construction observable from the harness. This was the residual the spike
  itself named sharpest: DRY is the one call signature the bump rewrites, and it had no positive
  control. Landed *before* the bump, as the spike recommended.
- **#1489** — the pin bump, with the `UD-Q4_K_XL` descriptor riding it, ADR-002 § "Pin Strategy"
  rewritten as durable policy, and the ADR-011 / `engine.md` sweeps this thread asked for.
  **ADR-011 P3–P5 PASSED on device 2026-08-15**; `gemma4E2BQAT` is now both
  `defaultInitialModelID` and `recommendedModelID`.
- **#1491** — the drift #1489 left behind, below.

### What #1491 found, since it changed a measurement this issue produced

#1489's final reviewer flagged an internal contradiction in ADR-002 bar item 2 and left it out of
scope. Resolved by **measuring** instead of re-reading — both tags fetched and `diff -u`'d:

- **`rbudget` was never new.** b8694 applies it ahead of the grammar in both passes too, inside a
  region byte-identical across the tags. "Upstream *now* applies…" was the false half; Pastura's
  missing counterpart is a **pre-existing** divergence.
- **The changed region was understated** — two hunks, not one: the backend-sampler early return's
  internals *and* the `set_logits` hoist above that block. Pastura reaches neither.
- **The mechanism**: bar item 2 asked only that the upstream body be *re-read at the new tag*, so a
  stage seen for the first time there read as new. It now requires a **cross-tag diff** — the two
  sibling bullets under the same bar, which did diff, both came out right the first time.

Also swept: `.claude/rules/engine.md`'s mirror of the same claim, three upstream citations that bar
item 6 requires to be greppable and were not, and `eval-log.md` §2026-08-08's two P3–P5 statements
(one born wrong in #1489 itself, one true-when-written and merely ambiguous — separated by `blame`,
and fixed with different instruments).

### Still open, deliberately

- **#1416** — QAT-Mobile `UD-Q2_K_XL`, blocked upstream on Metal `TQ`-family kernels. Not actionable
  here.
- **Device-only risk beyond P3–P5** — binary size and the ~8 GB RAM headroom estimate (ADR-002 §4)
  remain Mac-harness-invisible; the P3–P5 PASS licenses adoption, **not** a `minRAM` change.
- **`append_eval.py` still has no pin or GGUF dimension** in its `(date, profile_id)` key. The
  throwaway-`ModelProfile` workaround is recorded in the 2026-08-13 comment; nothing enforces it.
